### PR TITLE
KAFKA-3418: add javadoc section describing consumer failure detection

### PR DESCRIPTION
**Pull request by hachikuji**

### CODE REVIEW

**Comment by hachikuji** [comment]
cc @jkreps

**Comment by omkreddy** [comment]
LGTM

File: clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java [hunk 1]
```diff
@@ -128,6 +129,48 @@
...
+ *     a rebalance can be completed. However, clean shutdown with {@link #close()} is not impacted since
+ *     the consumer will send an explicit message to the server to leave the group and cause an immediate
+ *     rebalance.</li>
+ *     <li><code>max.poll.records</code>: Processing time in the poll loop is typically proportional to the number
+ *     of records processed, so it's natural to want to set a limit on the number of records handled at once.
+ *     This setting provides that. By default, there is essentially no limit.</li>
+ * </ol>
+ * <p>
+ * For use cases where message processing time varies unpredictably, neither of these options may be viable.
+ * The recommended way to handle these cases is to move message processing to another thread, which allows
```
Comment (ijuma) [inline clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java:166]: Is it correct that another downside of this approach is that the foreground thread could end up buffering too much data as we have no way to apply backpressure?

Comment (ewencp) [inline clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java:166]: @ijuma I think this question is addressed below -- pausing partitions in the network thread effectively is backpressure and turns the network thread into a pure heartbeating thread while processing is being performed. You can also, of course, choose to buffer as much or as little as you want by adjusting when you decide to pause the collection.

I'd say the current docs explain this well enough, though I think a few code examples (in the somewhat defunct examples jar) would be the most helpful way to show how to make this work in practice.

**Comment by ewencp** [comment]
LGTM. We may want to move some of this into the normal docs given how large this part of the javadocs is getting, but better to get this in _somewhere_ for the next release.

File: clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java (hunk 1 above)
Comment (ijuma) [inline clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java:166]: Makes sense @ewencp, not sure how I missed that sentence when I read it originally.

**Pull request closed without merging**